File: apps/backend/internal/model/asset.go
package model

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// Asset represents a homelab asset (server, VM, container, etc.)
type Asset struct {
	ID        uuid.UUID       `json:"id" db:"id"`
	UserID    string          `json:"user_id" db:"user_id"`
	Name      string          `json:"name" db:"name"`
	Type      *string         `json:"type,omitempty" db:"type"`
	Hostname  *string         `json:"hostname,omitempty" db:"hostname"`
	Metadata  json.RawMessage `json:"metadata,omitempty" db:"metadata"`
	CreatedAt time.Time       `json:"created_at" db:"created_at"`
	UpdatedAt time.Time       `json:"updated_at" db:"updated_at"`
}

// CreateAssetRequest is the DTO for creating a new asset
type CreateAssetRequest struct {
	Name     string           `json:"name" validate:"required,max=100"`
	Type     *string          `json:"type,omitempty" validate:"omitempty,max=50"`
	Hostname *string          `json:"hostname,omitempty" validate:"omitempty,max=255"`
	Metadata *json.RawMessage `json:"metadata,omitempty"`
}

// UpdateAssetRequest is the DTO for updating an existing asset
type UpdateAssetRequest struct {
	Name     *string          `json:"name,omitempty" validate:"omitempty,max=100"`
	Type     *string          `json:"type,omitempty" validate:"omitempty,max=50"`
	Hostname *string          `json:"hostname,omitempty" validate:"omitempty,max=255"`
	Metadata *json.RawMessage `json:"metadata,omitempty"`
}

// AssetResponse is the DTO for single asset responses
type AssetResponse struct {
	ID        uuid.UUID       `json:"id"`
	UserID    string          `json:"user_id"`
	Name      string          `json:"name"`
	Type      *string         `json:"type,omitempty"`
	Hostname  *string         `json:"hostname,omitempty"`
	Metadata  json.RawMessage `json:"metadata,omitempty"`
	CreatedAt time.Time       `json:"created_at"`
	UpdatedAt time.Time       `json:"updated_at"`
}

// NewAssetResponse converts an Asset domain model to AssetResponse DTO
func NewAssetResponse(asset *Asset) *AssetResponse {
	if asset == nil {
		return nil
	}

	return &AssetResponse{
		ID:        asset.ID,
		UserID:    asset.UserID,
		Name:      asset.Name,
		Type:      asset.Type,
		Hostname:  asset.Hostname,
		Metadata:  asset.Metadata,
		CreatedAt: asset.CreatedAt,
		UpdatedAt: asset.UpdatedAt,
	}
}

// AssetListResponse is the DTO for paginated asset list responses
type AssetListResponse struct {
	Assets []AssetResponse `json:"assets"`
	Total  int64           `json:"total"`
	Limit  int             `json:"limit"`
	Offset int             `json:"offset"`
}

// NewAssetListResponse converts a slice of Assets to AssetListResponse with pagination metadata
func NewAssetListResponse(assets []*Asset, total int64, limit, offset int) *AssetListResponse {
	responses := make([]AssetResponse, 0, len(assets))
	for _, asset := range assets {
		if resp := NewAssetResponse(asset); resp != nil {
			responses = append(responses, *resp)
		}
	}

	return &AssetListResponse{
		Assets: responses,
		Total:  total,
		Limit:  limit,
		Offset: offset,
	}
}

File: packages/zod/src/asset.ts
import { z } from "zod";
import { ZBase, ZUuid } from "./common.js";

/**
 * Asset Zod schemas matching Go models
 */

// Asset type enum
export const ZAssetType = z.enum(["server", "vm", "nas", "container", "network", "other"]);

// Asset metadata - flexible JSON structure for asset-specific details
// Examples: CPU specs, RAM, IP addresses, ports, etc.
export const ZAssetMetadata = z.record(z.any());

// Core Asset schema - matches Go model.Asset struct
export const ZAsset = ZBase.extend({
    user_id: z.string(),
    name: z.string().min(1).max(100),
    type: ZAssetType.nullable().optional(),
    hostname: z.string().max(255).nullable().optional(),
    metadata: ZAssetMetadata.nullable().optional(),
});

export type Asset = z.infer<typeof ZAsset>;

// Create Asset request - matches Go model.CreateAssetRequest
export const ZCreateAssetRequest = z.object({
    name: z.string().min(1).max(100),
    type: ZAssetType.optional(),
    hostname: z.string().max(255).optional(),
    metadata: ZAssetMetadata.optional(),
});

// Update Asset request - matches Go model.UpdateAssetRequest (all fields optional for PATCH)
export const ZUpdateAssetRequest = z.object({
    name: z.string().min(1).max(100).optional(),
    type: ZAssetType.optional(),
    hostname: z.string().max(255).optional(),
    metadata: ZAssetMetadata.optional(),
});

// Asset query parameters - matches Go model.AssetQueryParams
export const ZAssetQueryParams = z.object({
    limit: z.coerce.number().int().min(1).max(100).optional(),
    offset: z.coerce.number().int().min(0).optional(),
    type: ZAssetType.optional(),
    search: z.string().max(100).optional(),
    sort_by: z.enum(["name", "created_at", "updated_at"]).optional(),
    sort_order: z.enum(["asc", "desc"]).optional(),
});

// Asset list response - matches Go model.AssetListResponse
export const ZAssetListResponse = z.object({
    assets: z.array(ZAsset),
    total: z.number().int(),
    limit: z.number().int(),
    offset: z.number().int(),
});
